File: algorithms/1614. Maximum Nesting Depth of the Parentheses.cpp
#include <vector>
#include <string>
#include <algorithm>

class Stack {
    std::vector<char> arr;
public:
    void push(char x) {
        arr.push_back(x);
    }

    void pop() {
        if (!arr.empty()) arr.pop_back();
    }

    int size() const {
        return arr.size();
    }
};

class Solution {
public:
    int maxDepth(std::string s) {
        Stack st;
        int maxSize = 0;
        for (char ch : s) {
            if (ch == '(') {
                st.push(ch);
                maxSize = std::max(maxSize, st.size());
            } else if (ch == ')') {
                st.pop();
            }
        }
        return maxSize;
    }
};

File: algorithms/python/1614. Maximum Nesting Depth of the Parentheses.py
class Solution:
    class Stack:
        def __init__(self):
            self.arr = []

        def push(self, x):
            self.arr.append(x)

        def pop(self):
            if self.arr:
                return self.arr.pop()
            return None

        def size(self):
            return len(self.arr)

    def maxDepth(self, s: str) -> int:
        st = self.Stack()
        max_size = 0
        for ch in s:
            if ch == '(':
                st.push(ch)
                max_size = max(max_size, st.size())
            elif ch == ')':
                st.pop()
        return max_size
